provider/provider.go

```diff
@@ -0,0 +1,101 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"hash/crc32"
+
+	"github.com/alphagov/paas-aiven-broker/provider/aiven"
+	"github.com/pivotal-cf/brokerapi"
+)
+
+const AIVEN_BASE_URL string = "https://api.aiven.io"
+const SERVICE_TYPE string = "elasticsearch"
+
+type AivenProvider struct {
+	Client aiven.Client
+	Config *Config
+}
+
+func New(configJSON []byte) (*AivenProvider, error) {
+	config, err := DecodeConfig(configJSON)
+	if err != nil {
+		return nil, err
+	}
+	client := aiven.NewHttpClient(AIVEN_BASE_URL, config.APIToken, config.Project)
+	return &AivenProvider{
+		Client: client,
+		Config: config,
+	}, nil
+}
+
+func (ap *AivenProvider) Provision(ctx context.Context, provisionData ProvisionData) (dashboardURL, operationData string, err error) {
+	plan, err := ap.Config.FindPlan(provisionData.Service.ID, provisionData.Plan.ID)
+	if err != nil {
+		return "", "", err
+	}
+	createServiceInput := &aiven.CreateServiceInput{
+		Cloud:       ap.Config.Cloud,
+		Plan:        plan.AivenPlan,
+		ServiceName: BuildServiceName(ap.Config.ServiceNamePrefix, provisionData.InstanceID),
+		ServiceType: SERVICE_TYPE,
+		UserConfig: aiven.UserConfig{
+			ElasticsearchVersion: plan.ElasticsearchVersion,
+		},
+	}
+	_, err = ap.Client.CreateService(createServiceInput)
+	return dashboardURL, operationData, err
+}
+
+func (ap *AivenProvider) Deprovision(ctx context.Context, deprovisionData DeprovisionData) (operationData string, err error) {
+	_, err = ap.Client.DeleteService(&aiven.DeleteServiceInput{
+		ServiceName: BuildServiceName(ap.Config.ServiceNamePrefix, deprovisionData.InstanceID),
+	})
+	if err != nil {
+		return "", err
+	}
+	return "", nil
+}
+
+func (ap *AivenProvider) Bind(ctx context.Context, bindData BindData) (binding brokerapi.Binding, err error) {
+	return brokerapi.Binding{}, fmt.Errorf("not implemented")
+}
+
+func (ap *AivenProvider) Unbind(ctx context.Context, unbindData UnbindData) (err error) {
+	return fmt.Errorf("not implemented")
+}
+
+func (ap *AivenProvider) Update(ctx context.Context, updateData UpdateData) (operationData string, err error) {
+	return "", fmt.Errorf("not implemented")
+}
+
+func (ap *AivenProvider) LastOperation(ctx context.Context, lastOperationData LastOperationData) (state brokerapi.LastOperationState, description string, err error) {
+	status, err := ap.Client.GetServiceStatus(&aiven.GetServiceStatusInput{
+		ServiceName: BuildServiceName(ap.Config.ServiceNamePrefix, lastOperationData.InstanceID),
+	})
+	if err != nil {
+		return "", "", err
+	}
+	lastOperationState, description := ProviderStatesMapping(status)
+	return lastOperationState, description, nil
+}
+
+func BuildServiceName(prefix, guid string) string {
+	checksum := crc32.ChecksumIEEE([]byte(guid))
+	return fmt.Sprintf("%s-%x", prefix, checksum)
```

Comment:
I am starting to doubt whether we should have added this hyphen. It means the prefix is limited to 7 characters. If it were 8 we could use the `DEPLOY_ENV` as the prefix and it fits with [the current `DEPLOY_ENV` limit in place](https://github.com/alphagov/paas-cf/blob/12b5c07315d02d27bdecee8473743bb2c29f7e6e/Makefile#L7)

Reply:
I agree. The 8 char limit on `DEPLOY_ENV` is already somewhat restrictive, and making it hard to pick suitable names for the London deployments. I'd be keen not to have to reduce this further.